### PR TITLE
O4 data collection

### LINKED ISSUE #43: Remove deepdish dependence
`gwinferno.preprocess.data_collection` makes use of deepdish to save processed pe/injections files. We should switch to saving these files in the form of x-arrays using `gwinferno.utils.DataSet'. Also need to make a helper function to save MCMC output in this format as well.

### LINKED ISSUE #56: Add support for o4a semi-analytic injections in `selection.py`
Currently, `get_injection_dict()` in `preprocess.selection` is only designed to handle injections through o3. 

- Note some priors were updated in o4, so check these as well and update as needed.
- consider removing `get_semianalytic_injection_dict' or replacing with a function that can take a generic set of injections (in case a user is not using the standard R&P catalog injection sets.

### LINKED ISSUE #57: Use Planck 2015 Cosmology everywhere
R&P analyst guidelines specify Planck 2015 as the standard cosmology.

Locations where Planck 2018 is currently used:

https://github.com/FarrOutLab/GWInferno/blob/117919f22dd09b51cd901cf9d84c22d486f987cd/gwinferno/cosmology.py#L24-L28

https://github.com/FarrOutLab/GWInferno/blob/500bdd1b0aaf73d66e84c1ade72abaaf191124c6/gwinferno/preprocess/data_collection.py#L15

https://github.com/FarrOutLab/GWInferno/blob/117919f22dd09b51cd901cf9d84c22d486f987cd/tests/cosmology_test.py#L9

### LINKED ISSUE #69: Add support for o4a PE data in `data_collection.py`

### LINKED ISSUE #74: Use different dL_to_z prior for different catalogs in `data_collection.py`

### PR DESCRIPTION
**Pull request by jaxengodfrey**
Closes #56 
Closes #69
Closes #57
Closes #43
Closes #74

### CODE REVIEW

**Comment by jaxengodfrey** [comment]
Need to fix unit tests for new functions before merging

File: gwinferno/cosmology.py [hunk 1]
```diff
@@ -45,10 +53,25 @@ def __init__(self, Ho, omega_matter, omega_radiation, omega_lambda, distance_uni
         self.OmegaLambda = omega_lambda
         self.OmegaKappa = 1.0 - (self.OmegaMatter + self.OmegaRadiation + self.OmegaLambda)
         assert self.OmegaKappa == 0, "we only implement flat cosmologies! OmegaKappa must be 0"
-        self.z = jnp.array([0.0])
-        self.Dc = jnp.array([0.0])
-        self.Vc = jnp.array([0.0])
-        self.extend(max_z=2.3, dz=DEFAULT_DZ)
+
+        if not os.path.exists(gwp + f"_{distance_unit}.h5"):
```
Comment (bfarr) [inline gwinferno/cosmology.py:57]: I think this caching scheme as it's written is _really_ prone to bugs.  From what I can see, the only thing that would trigger a rewriting of the cache file is a change `distance_unit`, while in practice there are several things that the user could change where they would expect a new cache to be written.

The whole while loop method of extending the redshift / comoving volume / ... seems like overkill... Don't we almost exclusively use a maximum redshift?  It seems like we could do this without a need for loops.

File: gwinferno/cosmology.py [hunk 2]
```diff
@@ -105,13 +128,13 @@ def dDcdz(self, z, mpc=False):
             return dDc / MPC_CGS
         return dDc
 
-    def dVcdz(self, z, Dc=None, dz=DEFAULT_DZ):
+    def dVcdz(self, z, Dc=None, dz=DEFAULT_DZ, Mpc=False):
```
[the remark this answers is not in the document]
Comment (jaxengodfrey) [reply gwinferno/cosmology.py]: I made Mpc the default unit in [92431e7](https://github.com/FarrOutLab/GWInferno/pull/70/commits/92431e70ca8147b65520aedff9b46188bed72518)

File: gwinferno/preprocess/selection.py [hunk 3]
```diff
@@ -5,16 +5,56 @@
...
-from .priors import joint_prior_from_isotropic_spins
 
+def get_o4a_cumulative_injection_dict(file, param_names, ifar=1, snr=10):
+    with h5py.File(file, "r") as ff:
+        total_generated = ff.attrs["total_generated"]
+        analysis_time = ff.attrs["analysis_time"]
+        injections = np.asarray(ff["events"][:])
 
-def get_injection_dict(fi, ifar=1, snr=11, spin=False, additional_cuts=None):
+    found = (injections["far_gstlal"] <= 1 / ifar) | (injections["semianalytic_observed_phase_maximized_snr_net"] >= snr)
```
[the remark this answers is not in the document]
Comment (jaxengodfrey) [reply gwinferno/preprocess/selection.py]: Fixed this as well in [92431e7](https://github.com/FarrOutLab/GWInferno/pull/70/commits/92431e70ca8147b65520aedff9b46188bed72518). Now all far from all pipelines are used.

File: gwinferno/cosmology.py (hunk 1 above)
Comment (jaxengodfrey) [reply gwinferno/cosmology.py]: Removed this and was able to significantly speed up integral calculations by jaxify-ing them in [92431e7](https://github.com/FarrOutLab/GWInferno/pull/70/commits/92431e70ca8147b65520aedff9b46188bed72518).

**Review by bfarr — APPROVED** [review]

**Pull request merged**